### ISSUE TITLE
Invalid metric title

### ISSUE THREAD

**Issue by Battleroid** [alxrem/prometheus-logstash-exporter#3]
```
...
# HELP logstash_pipeline_plugins_filters_patterns_per_field_[syslog_message_json][dsn_mta] 
# TYPE logstash_pipeline_plugins_filters_patterns_per_field_[syslog_message_json][dsn_mta] gauge
logstash_pipeline_plugins_filters_patterns_per_field_[syslog_message_json][dsn_mta]{id="27bef9058fb3055233a8ed1f3dff77cf1e8f1a10-23",name="grok"} 1
# HELP logstash_pipeline_plugins_filters_patterns_per_field_[syslog_message_json][email] 
# TYPE logstash_pipeline_plugins_filters_patterns_per_field_[syslog_message_json][email] gauge
logstash_pipeline_plugins_filters_patterns_per_field_[syslog_message_json][email]{id="27bef9058fb3055233a8ed1f3dff77cf1e8f1a10-20",name="grok"} 1
# HELP logstash_pipeline_plugins_filters_patterns_per_field_[syslog_message_json][recipient] 
# TYPE logstash_pipeline_plugins_filters_patterns_per_field_[syslog_message_json][recipient] gauge
logstash_pipeline_plugins_filters_patterns_per_field_[syslog_message_json][recipient]{id="27bef9058fb3055233a8ed1f3dff77cf1e8f1a10-21",name="grok"} 1
# HELP logstash_pipeline_plugins_filters_patterns_per_field_[syslog_message_json][relay] 
# TYPE logstash_pipeline_plugins_filters_patterns_per_field_[syslog_message_json][relay] gauge
logstash_pipeline_plugins_filters_patterns_per_field_[syslog_message_json][relay]{id="27bef9058fb3055233a8ed1f3dff77cf1e8f1a10-22",name="grok"} 1
...
```

Prometheus definitely doesn't like square brackets within metric names.

```
$ cat metrics | promtool check metrics
error while linting: text format parsing error in line 358: invalid metric name in comment
```

**Comment by alxrem** [comment]
May you tell me version of logstash and show part of config, responsible for the brackets in the names?

**Comment by Battleroid** [comment]
Sure thing.

This was from Logstash 5.6.7, the culprit was this portion of the config.

```
if [syslog_message_json][dsn_mta] {
  grok {
    match => ["[syslog_message_json][dsn_mta]", "(?<dsn_mta_domain>[^.]*\.[^.]{2,3}(?:\.[^.]{2,3})?$)"]
    tag_on_failure => ["dsn_mta_grokparsefailure"]
  }
}
```

I've tested this locally with 5.6.4 (looks like this could affect 5.6 and later) and something as benign as: 

```
input {
  file {
    path => ["events*.json"]
    sincedb_path => "/dev/null"
    start_position => "beginning"
    codec => "json_lines"
  }
}

filter {
  if [event][else] {
    grok {
      match => ["[event][something]", "(?<something>.*)"]
    }
  }
}

output {
  stdout {
    codec => "rubydebug"
  }
}
```

Will result in the following output on the pipeline stats:

```
{
  "host": "<redacted>",
  "version": "5.6.4",
  "http_address": "127.0.0.1:9600",
  "id": "a3a26896-b548-44bc-bce4-5b0500d5794e",
  "name": "<redacted>",
  "pipeline": {
    "events": {
      "duration_in_millis": 0,
      "out": 0,
      "filtered": 0
    },
    "plugins": {
      "inputs": [
        {
          "id": "78f5d846da4d1008936b1bb2a226c14c5e45d58a-1",
          "name": "file"
        }
      ],
      "filters": [
        {
          "id": "78f5d846da4d1008936b1bb2a226c14c5e45d58a-2",
          "events": {
            "duration_in_millis": 0,
            "in": 0,
            "out": 0
          },
          "patterns_per_field": {
            "[event][something]": 1
          },
          "name": "grok"
        }
      ],
      "outputs": [
        {
          "id": "78f5d846da4d1008936b1bb2a226c14c5e45d58a-3",
          "events": {
            "duration_in_millis": 0,
            "in": 0,
            "out": 0
          },
          "name": "stdout"
        }
      ]
    },
    "reloads": {
      "last_error": null,
      "successes": 0,
      "last_success_timestamp": null,
      "last_failure_timestamp": null,
      "failures": 1
    },
    "queue": {
      "type": "memory"
    },
    "id": "main"
  }
}
```

Which then becomes:

```
...
# HELP logstash_pipeline_plugins_filters_patterns_per_field_[event][something] 
# TYPE logstash_pipeline_plugins_filters_patterns_per_field_[event][something] gauge
logstash_pipeline_plugins_filters_patterns_per_field_[event][something]{id="78f5d846da4d1008936b1bb2a226c14c5e45d58a-2",name="grok"} 1
...
```

**Comment by Battleroid** [comment]
I've done something stupid that for now at least works. 🤷‍♂️ 

```diff
diff --git a/main.go b/main.go
index 4b3ddce..571ef95 100644
--- a/main.go
+++ b/main.go
@@ -83,11 +83,12 @@ func (e *Exporter) collectMetrics(stats *Stats, ch chan<- prometheus.Metric) {
 }
 
 func (e *Exporter) collectTree(name string, data interface{}, labels prometheus.Labels, ch chan<- prometheus.Metric) {
+       metricName := strings.Replace(strings.Replace(strings.TrimRight(name, "]"), "_[", "__", 1), "][", "__", -1)
        if v, ok := data.(float64); ok {
                if len(labels) == 0 {
                        metric := prometheus.NewGauge(prometheus.GaugeOpts{
                                Namespace: namespace,
-                               Name:      name,
+                               Name:      metricName,
                        })
                        metric.Set(v)
                        ch <- metric
@@ -98,7 +99,7 @@ func (e *Exporter) collectTree(name string, data interface{}, labels prometheus.
                        }
                        vec := prometheus.NewGaugeVec(prometheus.GaugeOpts{
                                Namespace: namespace,
-                               Name:      name,
+                               Name:      metricName,
                        }, labelNames)
                        vec.With(labels).Set(v)
                        vec.Collect(ch)
```

Which gives me:

```
# HELP logstash_pipeline_plugins_filters_patterns_per_field__event__something 
# TYPE logstash_pipeline_plugins_filters_patterns_per_field__event__something gauge
logstash_pipeline_plugins_filters_patterns_per_field__event__something{id="ecc97b0317aa6db1f5394ae56bdb17561447fca7-2",name="grok"} 1
```

**Comment by alxrem** [comment]
In this case, the field name must not be part of the metric name. Now the field name is the label of the metric `pipeline_plugins_filters_patterns_per_field`, f.e.

    logstash_pipeline_plugins_filters_patterns_per_field{field="[event][something]",id="268ac2463828e600454039930ee02991c9e0391a30001b3e31729f1b234eb63f",name="grok",pipeline="main"} 1

This fix is released in the version 0.4.0.

Thank you for the bug report and examples.
